File: action/src/main/java/move/action/ScheduledActionManager.java
package move.action;

import com.google.common.base.Preconditions;
import com.google.common.util.concurrent.AbstractExecutionThreadService;
import com.google.common.util.concurrent.AbstractIdleService;
import com.google.common.util.concurrent.AbstractScheduledService;
import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.core.ILock;
import io.vertx.rxjava.core.Vertx;
import javaslang.control.Try;
import move.cluster.HazelcastProvider;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

/**
 *
 */
@Singleton
public class ScheduledActionManager extends AbstractIdleService {
    private static final Object EMPTY = new Object();
    private static final Logger LOG = LoggerFactory.getLogger(ScheduledActionManager.class);
    private final List<ClusterSingleton> clusterSingletons = new ArrayList<>();
    private final List<NodeSingleton> nodeSingletons = new ArrayList<>();
    @Inject
    Vertx vertx;
    @Inject
    HazelcastProvider hazelcastProvider;
    HazelcastInstance hazelcastInstance;

    @Inject
    ScheduledActionManager() {
    }

    @Override
    protected void startUp() throws Exception {
        hazelcastInstance = hazelcastProvider != null ? hazelcastProvider.get() : null;
        ActionManager.getScheduledActionMap().values().forEach(scheduledActionProvider -> {
            scheduledActionProvider.init();
            switch (scheduledActionProvider.getScheduledAction().type()) {
                case CLUSTER_SINGLETON:
                    clusterSingletons.add(new ClusterSingleton(scheduledActionProvider));
                    break;
                case NODE_SINGLETON:
                    nodeSingletons.add(new NodeSingleton(scheduledActionProvider));
                    break;
            }
        });

        nodeSingletons.forEach(value -> value.startAsync().awaitRunning());
        clusterSingletons.forEach(value -> value.startAsync().awaitRunning());
    }

    @Override
    protected void shutDown() throws Exception {
        clusterSingletons.forEach(value -> value.stopAsync().awaitTerminated());
        nodeSingletons.forEach(value -> value.stopAsync().awaitTerminated());
    }

    /**
     *
     */
    private class ClusterSingleton extends AbstractExecutionThreadService {
        private final ScheduledActionProvider<?> provider;
        private final int intervalSeconds;
        private Thread thread;

        public ClusterSingleton(ScheduledActionProvider<?> provider) {
            this.provider = provider;
            this.intervalSeconds = provider.getScheduledAction().intervalSeconds();

            Preconditions.checkNotNull(intervalSeconds, "ScheduledAction: " +
                provider.getActionClass().getCanonicalName() +
                " has an invalid value for intervalSeconds() = " +
                intervalSeconds);
        }

        @Override
        protected String serviceName() {
            return provider.getActionClass().getCanonicalName();
        }

        @Override
        protected void startUp() throws Exception {
        }

        @Override
        protected void triggerShutdown() {
            Try.run(() -> thread.interrupt());
        }

        @Override
        protected void run() throws Exception {
            thread = Thread.currentThread();
            while (isRunning() && !Thread.interrupted()) {
                try {
                    if (hazelcastInstance == null) {
                        try {
                            while (isRunning()) {
                                doRun();
                            }
                        } catch (InterruptedException e) {
                            return;
                        } catch (Exception e) {
                            LOG.error("Error Running Standalone Scheduled Action  " + provider.getActionClass().getCanonicalName(), e);
                            return;
                        }
                    } else {
                        ILock lock = null;

                        try {
                            lock = hazelcastInstance.getLock(provider.getActionClass().getCanonicalName());
                            lock.lockInterruptibly();
                        } catch (InterruptedException e) {
                            return;
                        } catch (Exception e1) {
                            LOG.info("Failed to get cluster lock for Scheduled Action " + provider.getActionClass().getCanonicalName(), e1);
                        }

                        try {
                            while (isRunning() && !Thread.interrupted()) {
                                try {
                                    doRun();
                                } catch (InterruptedException e) {
                                    return;
                                } catch (Exception e1) {
                                    LOG.error("Error Running Hazelcast Scheduled Action " + provider.getActionClass().getCanonicalName(), e1);
                                }
                            }
                        } finally {
                            assert lock != null;
                            lock.unlock();
                        }
                    }
                } catch (Throwable e) {
                    // Ignore.
                    LOG.warn("Failed to run Scheduled Action " + provider.getActionClass().getCanonicalName(), e);
                }
            }
        }

        private void doRun() throws InterruptedException {
            final long start = System.currentTimeMillis();
            provider.observe(EMPTY).toBlocking().first();

            final long elapsed = System.currentTimeMillis() - start;
            final long sleepFor = TimeUnit.SECONDS.toMillis(intervalSeconds) - elapsed;

            try {
                if (sleepFor > 0)
                    Thread.sleep(sleepFor);
            } catch (InterruptedException e) {
                // Do nothing.
                throw e;
            }
        }
    }

    /**
     *
     */
    private class NodeSingleton extends AbstractScheduledService {
        private final ScheduledActionProvider<?> provider;

        public NodeSingleton(ScheduledActionProvider<?> provider) {
            this.provider = provider;
        }

        @Override
        protected String serviceName() {
            return provider.getActionClass().getCanonicalName();
        }

        @Override
        protected void runOneIteration() throws Exception {
            try {
                run();
            } catch (InterruptedException e) {
                LOG.warn(provider.getActionClass().getCanonicalName(), e);
                throw e;
            } catch (Throwable e) {
                LOG.warn(provider.getActionClass().getCanonicalName(), e);
            }
        }

        protected void run() throws InterruptedException {
            provider.observe(EMPTY).toBlocking().first();
        }

        @Override
        protected Scheduler scheduler() {
            return Scheduler.newFixedRateSchedule(
                0,
                provider.getScheduledAction().intervalSeconds(),
                TimeUnit.SECONDS
            );
        }
    }
}

File: action/src/main/java/move/action/ScheduledActionManager.kt
package move.action

import com.google.common.base.Preconditions
import com.google.common.util.concurrent.AbstractExecutionThreadService
import com.google.common.util.concurrent.AbstractIdleService
import com.google.common.util.concurrent.AbstractScheduledService
import com.hazelcast.core.HazelcastInstance
import com.hazelcast.core.ILock
import io.vertx.rxjava.core.Vertx
import javaslang.control.Try
import move.cluster.HazelcastProvider
import org.slf4j.LoggerFactory
import java.util.*
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

/**

 */
@Singleton
class ScheduledActionManager @Inject
internal constructor(val vertx: Vertx,
                     hazelcastProvider: HazelcastProvider?) : AbstractIdleService() {
    private val clusterSingletons = ArrayList<ClusterSingleton>()
    private val nodeSingletons = ArrayList<NodeSingleton>()

    internal var hazelcastInstance: HazelcastInstance? = hazelcastProvider?.get()
    @Throws(Exception::class)
    override fun startUp() {
        ActionManager.scheduledActionMap.values.forEach { scheduledActionProvider ->
            when (scheduledActionProvider.scheduledAction.type) {
                ScheduledActionType.CLUSTER_SINGLETON -> clusterSingletons.add(ClusterSingleton(scheduledActionProvider))
                ScheduledActionType.NODE_SINGLETON -> nodeSingletons.add(NodeSingleton(scheduledActionProvider))
            }
        }

        nodeSingletons.forEach { value -> value.startAsync().awaitRunning() }
        clusterSingletons.forEach { value -> value.startAsync().awaitRunning() }
    }

    @Throws(Exception::class)
    override fun shutDown() {
        clusterSingletons.forEach { value -> value.stopAsync().awaitTerminated() }
        nodeSingletons.forEach { value -> value.stopAsync().awaitTerminated() }
    }

    /**

     */
    private inner class ClusterSingleton(private val provider: ScheduledActionProvider<*>) : AbstractExecutionThreadService() {
        private val intervalSeconds: Int
        private var thread: Thread? = null
        private var startup = true

        init {
            this.intervalSeconds = provider.scheduledAction.intervalSeconds

            Preconditions.checkNotNull(intervalSeconds, "ScheduledAction: " +
                    provider.actionClass.canonicalName +
                    " has an invalid value for intervalSeconds() = " +
                    intervalSeconds)
        }

        override fun serviceName(): String {
            return provider.actionClass.canonicalName
        }

        @Throws(Exception::class)
        override fun startUp() {
        }

        override fun triggerShutdown() {
            Try.run { thread!!.interrupt() }
        }

        @Throws(Exception::class)
        override fun run() {
            thread = Thread.currentThread()
            while (isRunning && !Thread.interrupted()) {
                try {
                    if (hazelcastInstance == null) {
                        try {
                            while (isRunning) {
                                doRun()
                            }
                        } catch (e: InterruptedException) {
                            return
                        } catch (e: Exception) {
                            LOG.error("Error Running Standalone Scheduled Action  " + provider.actionClass.canonicalName, e)
                            return
                        }

                    } else {
                        var lock: ILock? = null

                        try {
                            lock = hazelcastInstance!!.getLock(provider.actionClass.canonicalName)
                            lock.lockInterruptibly()
                        } catch (e: InterruptedException) {
                            return
                        } catch (e1: Exception) {
                            LOG.info("Failed to get cluster lock for Scheduled Action " + provider.actionClass.canonicalName, e1)
                        }

                        try {
                            while (isRunning && !Thread.interrupted()) {
                                try {
                                    doRun()
                                } catch (e: InterruptedException) {
                                    return
                                } catch (e1: Exception) {
                                    LOG.error("Error Running Hazelcast Scheduled Action " + provider.actionClass.canonicalName, e1)
                                }

                            }
                        } finally {
                            assert(lock != null)
                            lock!!.unlock()
                        }
                    }
                } catch (e: Throwable) {
                    // Ignore.
                    LOG.warn("Failed to run Scheduled Action " + provider.actionClass.canonicalName, e)
                }

            }
        }

        @Throws(InterruptedException::class)
        private fun doRun() {
            val start = System.currentTimeMillis()

            if (startup) {
                startup = false
            } else {
                provider.blockingBuilder(Unit)
            }

            val elapsed = System.currentTimeMillis() - start
            val sleepFor = TimeUnit.SECONDS.toMillis(intervalSeconds.toLong()) - elapsed

            try {
                if (sleepFor > 0) {
                    Thread.sleep(sleepFor)
                }
            } catch (e: InterruptedException) {
                // Do nothing.
                throw e
            }

        }
    }

    /**

     */
    private inner class NodeSingleton(private val provider: ScheduledActionProvider<*>) : AbstractScheduledService() {
        private var startup = true

        override fun serviceName(): String {
            return provider.actionClass.canonicalName
        }

        @Throws(Exception::class)
        override fun runOneIteration() {
            try {
                run()
            } catch (e: InterruptedException) {
                LOG.warn(provider.actionClass.canonicalName, e)
                throw e
            } catch (e: Throwable) {
                LOG.warn(provider.actionClass.canonicalName, e)
            }
        }

        @Throws(InterruptedException::class)
        protected fun run() {
            if (startup) {
                startup = false
            } else {
                provider.blockingBuilder(Unit)
            }
        }

        override fun scheduler(): AbstractScheduledService.Scheduler {
            return AbstractScheduledService.Scheduler.newFixedRateSchedule(
                    0,
                    provider.scheduledAction.intervalSeconds.toLong(),
                    TimeUnit.SECONDS
            )
        }
    }

    companion object {
        private val EMPTY = Any()
        private val LOG = LoggerFactory.getLogger(ScheduledActionManager::class.java)
    }
}
